src/ripple/overlay/Slot.h

```diff
@@ -0,0 +1,607 @@
+ * in the following states: 1) Counting. This is the peer selection state
+ * when Slot counts the messages and selects the peers; 2) Selected. Slot
+ * doesn't count messages in this state but a received message
+ * may transition Slot to Counting state.
+ */
+template <typename Peer, typename clock_type>
+class Slot final
+{
+private:
+    friend class Slots<Peer, clock_type>;
+    using id_t = typename Peer::id_t;
+    using time_point = typename clock_type::time_point;
+
+    Slot() : lastSelected_(clock_type::now()), state_(SlotState::Counting)
+    {
+    }
+
+    /** Update peer info. If the message is from a new
+     * peer or from a previously expired squelched peer then switch
+     * the peer's and slot's state to Counting. If time of last
+     * selection round is > MAX_UNSQUELCH_EXPIRE then switch the slot's
+     * state to Counting. If the number of messages for the peer
+     * is greater than MESSAGE_THRESHOLD then set the peer's
+     * state to Selected. If the number of selected peers is >= 75%
+     * of MAX_PEERS then switch slot's state to Selected,
+     * randomly select MAX_PEERS from selected peers, and call f() for
+     * each peer, which is not selected and not already in Squelched state.
+     * Set the state for those peers to Squelched and reset the count of
+     * all peers. Set slot's state to Selected. Message count is not updated
+     * when the slot is in Selected state.
+     * @param id Peer id which received the message
+     * @param peerPtr Peer which received the message
+     * @param type  Message type (Validation and Propose Set only,
+     *     others are ignored, future use)
+     * @param f Function is called for every peer that has to be
+     *     squelched with arguments:
+     *     peer weak_ptr
+     *     squelch duration
+     */
+    template <typename F>
+    void
+    update(
+        const id_t& id,
+        std::weak_ptr<Peer> peerPtr,
+        protocol::MessageType type,
+        F&& f);
+
+    /** Handle peer deletion when a peer disconnects.
+     * If the peer is in Selected state then
+     * call f() for every peer in squelched state and reset every
+     * peer's state to Counting. Switch Slot's state to Counting.
+     * @param id Deleted peer id
+     * @param erase If true then erase the peer. The peer is not erased
+     *      when the peer when is idled. The peer is deleted when it
+     *      disconnects
+     * @param f Function is called for every peer in Squelched state
+     *     with arguments:
+     *     peer weak_ptr
+     */
+    template <typename F>
+    void
+    deletePeer(id_t const& id, bool erase, F&& f);
+
+    /** Get the time of the last peer selection round */
+    const time_point&
+    getLastSelected() const
+    {
+        return lastSelected_;
+    }
+
+    /** Return number of peers in state/not in state */
+    template <typename Comp = std::equal_to<>>
+    std::uint16_t
+    inState(PeerState state, Comp comp = {}) const;
```

Comment:
The `comp` parameter here is a consequence of making this class generic. If we could solve that we could get rid of this parameter as well (here and below).